src/ape_cache/query.py

```diff
@@ -0,0 +1,337 @@
+            FROM contract_events
+            WHERE contract_events.block_number >= :start_block
+            AND contract_events.block_number <= :stop_block
+            AND contract_events.block_number mod :step = 0
+        """
+
+    @singledispatchmethod
+    def perform_query_stmt(self, query: QueryType) -> str:
+        raise QueryEngineError("Not a compatible QueryType")
+
+    @perform_query_stmt.register
+    def perform_block_stmt(self, query: BlockQuery) -> str:
+        return """
+            SELECT :columns
+            FROM blocks
+            WHERE blocks.number >= :start_block
+            AND blocks.number <= :stop_block
+            AND blocks.number mod :step = 0
+        """
+
+    @perform_query_stmt.register
+    def perform_transaction_stmt(self, query: BlockTransactionQuery) -> str:
+        return """
+            SELECT :columns
+            FROM transactions
+            WHERE transactions.block_id = :block_id
+        """
+
+    @perform_query_stmt.register
+    def perform_contract_event_stmt(self, query: ContractEventQuery) -> str:
+        return """
+            SELECT :columns
+            FROM contract_events
+            WHERE contract_events.block_number >= :start_block
+            AND contract_events.block_number <= :stop_block
+            AND contract_events.block_number mod :step = 0
+        """
+
+    @singledispatchmethod
+    def estimate_query(self, query: QueryType) -> Optional[int]:  # type: ignore
+        return None  # can't handle this query
+
+    @estimate_query.register
+    def estimate_block_query(self, query: BlockQuery) -> Optional[int]:
+        try:
+            if (
+                self._block_query(self.estimate_query_stmt(query), query)
+                == (query.stop_block - query.start_block) // query.step
+            ):
+                # NOTE: Assume 200 msec to get data from database
+                return 200
+            # Can't handle this query
+            # TODO: Allow partial queries
+            return None
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
+            return None
+
+    @estimate_query.register
+    def estimate_block_transaction_query(self, query: BlockTransactionQuery) -> Optional[int]:
+        try:
+            with self.engine.connect() as conn:
+                q = conn.execute(
+                    text(self.estimate_query_stmt(query)),
+                    block_id=query.block_id,
+                )
+                if q.rowcount > 0:
+                    # NOTE: Assume 200 msec to get data from database
+                    return 200
+                # Can't handle this query
+                return None
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
```

Comment:
Would this be better at the `error` level instead of `debug` ?
(Regarding everywhere where we catch all exceptions and skip)

Reply:
I agree with this. Please change

Reply:
I agree with this. Please change